packages/rollup-contracts/contracts/ovm/PartialStateManager.sol

```diff
@@ -0,0 +1,329 @@
+
+/**
+ * @title PartialStateManager
+ * @notice The PartialStateManager is used for the on-chain fraud proof checker.
+ *         It is supplied with only the state which is used to execute a single transaction. This
+ *         is unlike the FullStateManager which has access to every storage slot.
+ */
+contract PartialStateManager {
+    address constant ZERO_ADDRESS = 0x0000000000000000000000000000000000000000;
+
+    SafetyChecker safetyChecker;
+    FraudVerifier fraudVerifier;
+    ExecutionManager executionManager;
+
+
+    mapping(address=>mapping(bytes32=>bytes32)) ovmContractStorage;
+    mapping(address=>uint) ovmContractNonces;
+    mapping(address=>address) ovmCodeContracts;
+
+    bool public existsInvalidStateAccess;
+    bool public isTransactionSuccessfullyExecuted;
+
+    mapping(address=>mapping(bytes32=>bool)) isVerifiedStorage;
+    mapping(address=>bool) isVerifiedContract;
+    mapping(uint=>bytes32) updatedStorage;
+    uint updatedStorageCounter;
+    mapping(uint=>address) updatedContracts;
+    uint updatedContractsCounter;
+    bytes32 stateRoot;
+
+    modifier preExecution {
+        require(!isTransactionSuccessfullyExecuted, "Function only callable before tx execution.");
+        _;
+    }
+
+    modifier postExecution {
+        require(isTransactionSuccessfullyExecuted, "Function only callable after tx execution.");
+        _;
+    }
+
+    /**
+     * @notice Construct a new FullStateManager with a specified safety checker.
+     */
+    constructor(bytes32 _stateRoot, address _safetyCheckerAddress, address _fraudVerifierAddress) public {
+        stateRoot = _stateRoot;
+        safetyChecker = SafetyChecker(_safetyCheckerAddress);
+        fraudVerifier = FraudVerifier(_fraudVerifierAddress);
+    }
+
+    /**
+     * @notice This is a seperate function because it allows us to first deploy the state manager,
+     * then deploy the execution manager (passing in the state manager address), and then set the execution manager
+     * address in the state manager. This is a bit ugly & probably should be thought through a bit more.
+     */
+    function setExecutionManager(address _executionManagerAddress) public {
+        require(msg.sender == address(fraudVerifier));
+        executionManager = ExecutionManager(_executionManagerAddress);
+    }
+
+    /**
+     * @notice Initialize a new transaction execution
+     */
+    function initNewTransactionExecution() external {
+        require(msg.sender == address(fraudVerifier));
+        existsInvalidStateAccess = false;
+        updatedStorageCounter = 0;
+        updatedContractsCounter = 0;
+    }
+
+    function setIsTransitionSuccessfullyExecuted(bool _isTransactionSuccessfullyExecuted) external {
+        require(msg.sender == address(fraudVerifier));
+        isTransactionSuccessfullyExecuted = _isTransactionSuccessfullyExecuted;
+    }
+
+    function ensureVerifiedStorage(address _ovmContractAddress, bytes32 _slot) private {
```

Comment:
Nit: these names feels weird because we don't actually ensure anything, we just flag if it's not.  Maybe `checkStorageAccess` and `checkContractAccess`?  idk